suite/tests/client-interface/stolen-reg.c

```diff
@@ -0,0 +1,218 @@
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/*
+ * API regression test for stolen register translation.
+ */
+
+#include "tools.h"
+#include "thread.h"
+#include <stdio.h>
+#include <setjmp.h>
+#include <signal.h>
+#include <stdint.h> /* uintptr_t */
+#include <ucontext.h>
+
+#define STOLEN_REG_SENTINEL 42
+
+/* We assume a single thread when these are used. */
+static SIGJMP_BUF mark;
+static int sigsegv_count;
+
+static volatile int thread_finished;
+
+static uintptr_t
+get_stolen_reg_val(void)
+{
+    uintptr_t val = 0;
+#if defined(AARCH64)
+    __asm__ __volatile__("str x28, %0\n\t" : "=m"(val) : :);
+#elif defined(ARM)
+    __asm__ __volatile__("str r10, %0\n\t" : "=m"(val) : :);
+#else
+#    error Unsupported arch
+#endif
+    return val;
+}
+
+/* This can't be a regular function since the stolen reg is callee-saved. */
+#if defined(AARCH64)
+#    define SET_STOLEN_REG_TO_SENTINEL() \
+        __asm__ __volatile__("mov x28, #" STRINGIFY(STOLEN_REG_SENTINEL) : : : "x28")
+#elif defined(ARM)
+#    define SET_STOLEN_REG_TO_SENTINEL() \
+        __asm__ __volatile__("mov r10, #" STRINGIFY(STOLEN_REG_SENTINEL) : : : "r10")
+#else
+#    error Unsupported arch
+#endif
+
+static void
+signal_handler(int sig, siginfo_t *siginfo, ucontext_t *ucxt)
+{
+    if (sig != SIGSEGV)
+        return;
+    print("Got SIGSEGV\n");
+    uintptr_t val = get_stolen_reg_val();
+    if (val != STOLEN_REG_SENTINEL) {
+        print("ERROR: Stolen register %d not preserved on handler entry: %d\n",
+              STOLEN_REG_SENTINEL, val);
+    }
+    sigcontext_t *sc = SIGCXT_FROM_UCXT(ucxt);
+    if (sc->IF_ARM_ELSE(SC_R10, SC_R28) != STOLEN_REG_SENTINEL) {
+        print("ERROR: Stolen register %d not preserved in signal context: %d\n",
+              STOLEN_REG_SENTINEL, sc->IF_ARM_ELSE(SC_R10, SC_R28));
+    }
+    if (sigsegv_count == 0) {
+        /* Allow re-execution to no longer fault. */
+        sc->SC_R0 = sc->SC_XSP;
+    } else {
+        SIGLONGJMP(mark, 1);
+    }
+}
+
+static uintptr_t
+cause_sigsegv(void)
+{
+    uintptr_t val = 0;
+    /* Generate SIGSEGV with a sentinel in the stolen reg.
+     * This precise instruction sequence is matched by the client.
```

Comment:
Maybe add that the client code is in suite/tests/client-interface/stolen-reg.dll.c.

Reply:
Every suite/tests/client-interface test has the app as foo.c and the cilent as foo.dll.c, so I would say it's not necessary in these tests, but I suppose an explicit comment doesn't hurt.  Done.